### PR TITLE
typeck: don't suggest inaccessible private fields

### LINKED ISSUE #76077: E0063 should not suggest adding inaccessible private fields
This error message could be more helpful:

```rust
pub mod foo {
    pub struct Foo {
        you_cant_use_this_field: bool,
    }
}

fn main() {
    foo::Foo {};
}
```

> error[E0063]: missing field `you_cant_use_this_field` in initializer of `foo::Foo`

It complains about lack of a private field in the initializer, but this problem can't be solved by adding the field. The implied solution is impossible, and will only lead to another dead-end error.

It would be better to explain in general that this struct can't be constructed from a literal syntax here, because it contains private fields. Bonus points for finding methods that return `Self` and suggesting them.

[related forum thread](https://users.rust-lang.org/t/very-new-to-rust-whats-the-missing-base-error/47999)

### PR DESCRIPTION
**Pull request by davidtwco**
Fixes #76077.

This PR adjusts the missing field diagnostic logic in typeck so that when none of the missing fields in a struct expr are accessible then the error is less confusing.

r? @estebank

### CODE REVIEW

File: src/test/ui/issues/issue-76077.rs [hunk 1]
```diff
@@ -0,0 +1,10 @@
+pub mod foo {
+    pub struct Foo {
+        you_cant_use_this_field: bool,
+    }
+}
+
+fn main() {
+    foo::Foo {};
+    //~^ ERROR cannot construct `Foo` with struct literal syntax due to inaccessible fields
+}
```
Comment (estebank) [inline src/test/ui/issues/issue-76077.rs:10]: Can we add a test for the following case?

```rust
pub mod foo {
    pub struct Foo {
        you_cant_use_this_field: bool,
    }
    impl Foo {
        pub fn new() -> Foo {
            Foo {
                you_cant_use_this_field: true,
            }
        }
    }
}

fn main() {
    let foo::Foo {} = foo::Foo::new();
}
```

**Review by estebank — APPROVED** [review]

File: src/test/ui/issues/issue-76077.rs (hunk 1 above)
Comment (davidtwco) [reply src/test/ui/issues/issue-76077.rs:10]: Added a test for this and implemented these changes for patterns.

File: compiler/rustc_typeck/src/check/expr.rs [hunk 2]
```diff
@@ -1241,42 +1241,21 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
-                remaining_fields.keys().map(|ident| ident.as_str()).collect::<Vec<_>>();
-
-            displayable_field_names.sort();
+            let no_accessible_remaining_fields = remaining_fields
+                .iter()
+                .filter(|(_, (_, field))| {
+                    field.vis.is_accessible_from(tcx.parent_module(expr_id).to_def_id(), tcx)
+                })
+                .count()
+                == 0;
```
Comment (estebank) [inline compiler/rustc_typeck/src/check/expr.rs:1250]: I would normally write this as `.next().is_none()`, but you don't need to change it :)

File: src/test/ui/issues/issue-76077-1.stderr [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+LL |     let foo::Foo { .. } = foo::Foo::default();
+   |                  ^^^^^^
+
+error: pattern requires `..` due to inaccessible fields
+  --> $DIR/issue-76077-1.rs:16:9
+   |
+LL |     let foo::Bar { visible } = foo::Bar::default();
+   |         ^^^^^^^^^^^^^^^^^^^^
+   |
+help: add a `..`
```
Comment (estebank) [inline src/test/ui/issues/issue-76077-1.stderr:18]: I would change the wording to something along the lines of "ignore the inaccessible and unused fields"

**Comment by estebank** [comment]
@bors r+

Approving. Feel free to change the wording of the message and reapprove, but otherwise we can do it at a later time.

File: compiler/rustc_typeck/src/check/expr.rs (hunk 2 above)
Comment (davidtwco) [reply compiler/rustc_typeck/src/check/expr.rs]: Fixed!

File: src/test/ui/issues/issue-76077-1.stderr (hunk 3 above)
Comment (davidtwco) [reply src/test/ui/issues/issue-76077-1.stderr]: Fixed!

**Pull request merged**